File: sample/kotlin/07_game_of_life_loop.kt
import kotlin.math.*


// 07: Sample that outputs Game of Life evolution as a GIF.

fun next_state(grid: MutableList<Any?>, w: Long, h: Long): MutableList<Any?> {
    var nxt: MutableList<Any?> = __pytra_as_list(mutableListOf<Any?>())
    val __step_0 = __pytra_int(1L)
    var y = __pytra_int(0L)
    while ((__step_0 >= 0L && y < __pytra_int(h)) || (__step_0 < 0L && y > __pytra_int(h))) {
        var row: MutableList<Any?> = __pytra_as_list(mutableListOf<Any?>())
        val __step_1 = __pytra_int(1L)
        var x = __pytra_int(0L)
        while ((__step_1 >= 0L && x < __pytra_int(w)) || (__step_1 < 0L && x > __pytra_int(w))) {
            var cnt: Long = __pytra_int(0L)
            val __step_2 = __pytra_int(1L)
            var dy = __pytra_int((-1L))
            while ((__step_2 >= 0L && dy < __pytra_int(2L)) || (__step_2 < 0L && dy > __pytra_int(2L))) {
                val __step_3 = __pytra_int(1L)
                var dx = __pytra_int((-1L))
                while ((__step_3 >= 0L && dx < __pytra_int(2L)) || (__step_3 < 0L && dx > __pytra_int(2L))) {
                    if (((__pytra_int(dx) != __pytra_int(0L)) || (__pytra_int(dy) != __pytra_int(0L)))) {
                        var nx: Long = __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(dx))) + __pytra_int(w))) % __pytra_int(w)))
                        var ny: Long = __pytra_int((__pytra_int((__pytra_int((__pytra_int(y) + __pytra_int(dy))) + __pytra_int(h))) % __pytra_int(h)))
                        cnt += __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, ny)), nx))
                    }
                    dx += __step_3
                }
                dy += __step_2
            }
            var alive: Long = __pytra_int(__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)))
            if (((__pytra_int(alive) == __pytra_int(1L)) && ((__pytra_int(cnt) == __pytra_int(2L)) || (__pytra_int(cnt) == __pytra_int(3L))))) {
                row = __pytra_as_list(row); row.add(1L)
            } else {
                if (((__pytra_int(alive) == __pytra_int(0L)) && (__pytra_int(cnt) == __pytra_int(3L)))) {
                    row = __pytra_as_list(row); row.add(1L)
                } else {
                    row = __pytra_as_list(row); row.add(0L)
                }
            }
            x += __step_1
        }
        nxt = __pytra_as_list(nxt); nxt.add(row)
        y += __step_0
    }
    return __pytra_as_list(nxt)
}

fun render(grid: MutableList<Any?>, w: Long, h: Long, cell: Long): MutableList<Any?> {
    var width: Long = __pytra_int((__pytra_int(w) * __pytra_int(cell)))
    var height: Long = __pytra_int((__pytra_int(h) * __pytra_int(cell)))
    var frame: MutableList<Any?> = __pytra_as_list(__pytra_bytearray((__pytra_int(width) * __pytra_int(height))))
    val __step_0 = __pytra_int(1L)
    var y = __pytra_int(0L)
    while ((__step_0 >= 0L && y < __pytra_int(h)) || (__step_0 < 0L && y > __pytra_int(h))) {
        val __step_1 = __pytra_int(1L)
        var x = __pytra_int(0L)
        while ((__step_1 >= 0L && x < __pytra_int(w)) || (__step_1 < 0L && x > __pytra_int(w))) {
            var v: Long = __pytra_int(__pytra_ifexp((__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)) != 0L), 255L, 0L))
            val __step_2 = __pytra_int(1L)
            var yy = __pytra_int(0L)
            while ((__step_2 >= 0L && yy < __pytra_int(cell)) || (__step_2 < 0L && yy > __pytra_int(cell))) {
                var base: Long = __pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int(y) * __pytra_int(cell))) + __pytra_int(yy))) * __pytra_int(width))) + __pytra_int((__pytra_int(x) * __pytra_int(cell)))))
                val __step_3 = __pytra_int(1L)
                var xx = __pytra_int(0L)
                while ((__step_3 >= 0L && xx < __pytra_int(cell)) || (__step_3 < 0L && xx > __pytra_int(cell))) {
                    __pytra_set_index(frame, (__pytra_int(base) + __pytra_int(xx)), v)
                    xx += __step_3
                }
                yy += __step_2
            }
            x += __step_1
        }
        y += __step_0
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

fun run_07_game_of_life_loop() {
    var w: Long = __pytra_int(144L)
    var h: Long = __pytra_int(108L)
    var cell: Long = __pytra_int(4L)
    var steps: Long = __pytra_int(105L)
    var out_path: String = __pytra_str("sample/out/07_game_of_life_loop.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var grid: MutableList<Any?> = __pytra_as_list(run { val __out = mutableListOf<Any?>(); val __step = __pytra_int(1L); var __lc_i = __pytra_int(0L); while ((__step >= 0L && __lc_i < __pytra_int(h)) || (__step < 0L && __lc_i > __pytra_int(h))) { __out.add(__pytra_list_repeat(0L, w)); __lc_i += __step }; __out })
    val __step_0 = __pytra_int(1L)
    var y = __pytra_int(0L)
    while ((__step_0 >= 0L && y < __pytra_int(h)) || (__step_0 < 0L && y > __pytra_int(h))) {
        val __step_1 = __pytra_int(1L)
        var x = __pytra_int(0L)
        while ((__step_1 >= 0L && x < __pytra_int(w)) || (__step_1 < 0L && x > __pytra_int(w))) {
            var noise: Long = __pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int(x) * __pytra_int(37L))) + __pytra_int((__pytra_int(y) * __pytra_int(73L))))) + __pytra_int((__pytra_int((__pytra_int(x) * __pytra_int(y))) % __pytra_int(19L))))) + __pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(y))) % __pytra_int(11L))))) % __pytra_int(97L)))
            if ((__pytra_int(noise) < __pytra_int(3L))) {
                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, 1L)
            }
            x += __step_1
        }
        y += __step_0
    }
    var glider: MutableList<Any?> = __pytra_as_list(mutableListOf(mutableListOf(0L, 1L, 0L), mutableListOf(0L, 0L, 1L), mutableListOf(1L, 1L, 1L)))
    var r_pentomino: MutableList<Any?> = __pytra_as_list(mutableListOf(mutableListOf(0L, 1L, 1L), mutableListOf(1L, 1L, 0L), mutableListOf(0L, 1L, 0L)))
    var lwss: MutableList<Any?> = __pytra_as_list(mutableListOf(mutableListOf(0L, 1L, 1L, 1L, 1L), mutableListOf(1L, 0L, 0L, 0L, 1L), mutableListOf(0L, 0L, 0L, 0L, 1L), mutableListOf(1L, 0L, 0L, 1L, 0L)))
    val __step_2 = __pytra_int(18L)
    var gy = __pytra_int(8L)
    while ((__step_2 >= 0L && gy < __pytra_int((__pytra_int(h) - __pytra_int(8L)))) || (__step_2 < 0L && gy > __pytra_int((__pytra_int(h) - __pytra_int(8L))))) {
        val __step_3 = __pytra_int(22L)
        var gx = __pytra_int(8L)
        while ((__step_3 >= 0L && gx < __pytra_int((__pytra_int(w) - __pytra_int(8L)))) || (__step_3 < 0L && gx > __pytra_int((__pytra_int(w) - __pytra_int(8L))))) {
            var kind: Long = __pytra_int((__pytra_int((__pytra_int((__pytra_int(gx) * __pytra_int(7L))) + __pytra_int((__pytra_int(gy) * __pytra_int(11L))))) % __pytra_int(3L)))
            if ((__pytra_int(kind) == __pytra_int(0L))) {
                var ph: Long = __pytra_int(__pytra_len(glider))
                val __step_4 = __pytra_int(1L)
                var py = __pytra_int(0L)
                while ((__step_4 >= 0L && py < __pytra_int(ph)) || (__step_4 < 0L && py > __pytra_int(ph))) {
                    var pw: Long = __pytra_int(__pytra_len(__pytra_as_list(__pytra_get_index(glider, py))))
                    val __step_5 = __pytra_int(1L)
                    var px = __pytra_int(0L)
                    while ((__step_5 >= 0L && px < __pytra_int(pw)) || (__step_5 < 0L && px > __pytra_int(pw))) {
                        if ((__pytra_int(__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(glider, py)), px))) == __pytra_int(1L))) {
                            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), 1L)
                        }
                        px += __step_5
                    }
                    py += __step_4
                }
            } else {
                if ((__pytra_int(kind) == __pytra_int(1L))) {
                    var ph: Long = __pytra_int(__pytra_len(r_pentomino))
                    val __step_6 = __pytra_int(1L)
                    var py = __pytra_int(0L)
                    while ((__step_6 >= 0L && py < __pytra_int(ph)) || (__step_6 < 0L && py > __pytra_int(ph))) {
                        var pw: Long = __pytra_int(__pytra_len(__pytra_as_list(__pytra_get_index(r_pentomino, py))))
                        val __step_7 = __pytra_int(1L)
                        var px = __pytra_int(0L)
                        while ((__step_7 >= 0L && px < __pytra_int(pw)) || (__step_7 < 0L && px > __pytra_int(pw))) {
                            if ((__pytra_int(__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(r_pentomino, py)), px))) == __pytra_int(1L))) {
                                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), 1L)
                            }
                            px += __step_7
                        }
                        py += __step_6
                    }
                } else {
                    var ph: Long = __pytra_int(__pytra_len(lwss))
                    val __step_8 = __pytra_int(1L)
                    var py = __pytra_int(0L)
                    while ((__step_8 >= 0L && py < __pytra_int(ph)) || (__step_8 < 0L && py > __pytra_int(ph))) {
                        var pw: Long = __pytra_int(__pytra_len(__pytra_as_list(__pytra_get_index(lwss, py))))
                        val __step_9 = __pytra_int(1L)
                        var px = __pytra_int(0L)
                        while ((__step_9 >= 0L && px < __pytra_int(pw)) || (__step_9 < 0L && px > __pytra_int(pw))) {
                            if ((__pytra_int(__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(lwss, py)), px))) == __pytra_int(1L))) {
                                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), 1L)
                            }
                            px += __step_9
                        }
                        py += __step_8
                    }
                }
            }
            gx += __step_3
        }
        gy += __step_2
    }
    var frames: MutableList<Any?> = __pytra_as_list(mutableListOf<Any?>())
    val __step_11 = __pytra_int(1L)
    var __loop_10 = __pytra_int(0L)
    while ((__step_11 >= 0L && __loop_10 < __pytra_int(steps)) || (__step_11 < 0L && __loop_10 > __pytra_int(steps))) {
        frames = __pytra_as_list(frames); frames.add(render(grid, w, h, cell))
        grid = __pytra_as_list(next_state(grid, w, h))
        __loop_10 += __step_11
    }
    __pytra_noop(out_path, (__pytra_int(w) * __pytra_int(cell)), (__pytra_int(h) * __pytra_int(cell)), frames, mutableListOf<Any?>())
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

fun main(args: Array<String>) {
    run_07_game_of_life_loop()
}

File: sample/swift/07_game_of_life_loop.swift
import Foundation


// 07: Sample that outputs Game of Life evolution as a GIF.

func next_state(grid: [Any], w: Int64, h: Int64) -> [Any] {
    var nxt: [Any] = __pytra_as_list([])
    var y = __pytra_int(Int64(0))
    while (y < __pytra_int(h)) {
        var row: [Any] = __pytra_as_list([])
        var x = __pytra_int(Int64(0))
        while (x < __pytra_int(w)) {
            var cnt: Int64 = Int64(0)
            var dy = __pytra_int(-Int64(1))
            while (dy < __pytra_int(Int64(2))) {
                var dx = __pytra_int(-Int64(1))
                while (dx < __pytra_int(Int64(2))) {
                    if ((__pytra_int(dx) != __pytra_int(Int64(0))) || (__pytra_int(dy) != __pytra_int(Int64(0)))) {
                        var nx: Int64 = (((x + dx) + w) % w)
                        var ny: Int64 = (((y + dy) + h) % h)
                        cnt += __pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, ny)), nx))
                    }
                    dx += 1
                }
                dy += 1
            }
            var alive: Int64 = __pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x))
            if ((__pytra_int(alive) == __pytra_int(Int64(1))) && ((__pytra_int(cnt) == __pytra_int(Int64(2))) || (__pytra_int(cnt) == __pytra_int(Int64(3))))) {
                row.append(Int64(1))
            } else {
                if ((__pytra_int(alive) == __pytra_int(Int64(0))) && (__pytra_int(cnt) == __pytra_int(Int64(3)))) {
                    row.append(Int64(1))
                } else {
                    row.append(Int64(0))
                }
            }
            x += 1
        }
        nxt.append(row)
        y += 1
    }
    return nxt
}

func render(grid: [Any], w: Int64, h: Int64, cell: Int64) -> [Any] {
    var width: Int64 = (w * cell)
    var height: Int64 = (h * cell)
    var frame: [Any] = __pytra_as_list(__pytra_bytearray((width * height)))
    var y = __pytra_int(Int64(0))
    while (y < __pytra_int(h)) {
        var x = __pytra_int(Int64(0))
        while (x < __pytra_int(w)) {
            var v: Int64 = __pytra_int(__pytra_ifexp((__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)) != 0), Int64(255), Int64(0)))
            var yy = __pytra_int(Int64(0))
            while (yy < __pytra_int(cell)) {
                var base: Int64 = ((((y * cell) + yy) * width) + (x * cell))
                var xx = __pytra_int(Int64(0))
                while (xx < __pytra_int(cell)) {
                    __pytra_setIndex(frame, (base + xx), v)
                    xx += 1
                }
                yy += 1
            }
            x += 1
        }
        y += 1
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

func run_07_game_of_life_loop() {
    var w: Int64 = Int64(144)
    var h: Int64 = Int64(108)
    var cell: Int64 = Int64(4)
    var steps: Int64 = Int64(105)
    var out_path: String = "sample/out/07_game_of_life_loop.gif"
    var start: Double = __pytra_perf_counter()
    var grid: [Any] = __pytra_as_list(({ () -> [Any] in var __out: [Any] = []; let __step = __pytra_int(Int64(1)); var __lc_i = __pytra_int(Int64(0)); while ((__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h))) { __out.append(__pytra_list_repeat(Int64(0), w)); __lc_i += __step }; return __out })())
    var y = __pytra_int(Int64(0))
    while (y < __pytra_int(h)) {
        var x = __pytra_int(Int64(0))
        while (x < __pytra_int(w)) {
            var noise: Int64 = (((((x * Int64(37)) + (y * Int64(73))) + ((x * y) % Int64(19))) + ((x + y) % Int64(11))) % Int64(97))
            if (__pytra_int(noise) < __pytra_int(Int64(3))) {
                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(1))
            }
            x += 1
        }
        y += 1
    }
    var glider: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(0)], [Int64(0), Int64(0), Int64(1)], [Int64(1), Int64(1), Int64(1)]])
    var r_pentomino: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(1)], [Int64(1), Int64(1), Int64(0)], [Int64(0), Int64(1), Int64(0)]])
    var lwss: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(1), Int64(1), Int64(1)], [Int64(1), Int64(0), Int64(0), Int64(0), Int64(1)], [Int64(0), Int64(0), Int64(0), Int64(0), Int64(1)], [Int64(1), Int64(0), Int64(0), Int64(1), Int64(0)]])
    var gy = __pytra_int(Int64(8))
    let __step_2 = __pytra_int(Int64(18))
    while ((__step_2 >= 0 && gy < __pytra_int(h - Int64(8))) || (__step_2 < 0 && gy > __pytra_int(h - Int64(8)))) {
        var gx = __pytra_int(Int64(8))
        let __step_3 = __pytra_int(Int64(22))
        while ((__step_3 >= 0 && gx < __pytra_int(w - Int64(8))) || (__step_3 < 0 && gx > __pytra_int(w - Int64(8)))) {
            var kind: Int64 = (((gx * Int64(7)) + (gy * Int64(11))) % Int64(3))
            if (__pytra_int(kind) == __pytra_int(Int64(0))) {
                var ph: Int64 = __pytra_len(glider)
                var py = __pytra_int(Int64(0))
                while (py < __pytra_int(ph)) {
                    var pw: Int64 = __pytra_len(__pytra_as_list(__pytra_getIndex(glider, py)))
                    var px = __pytra_int(Int64(0))
                    while (px < __pytra_int(pw)) {
                        if (__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(glider, py)), px)) == __pytra_int(Int64(1))) {
                            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, ((gy + py) % h))), ((gx + px) % w), Int64(1))
                        }
                        px += 1
                    }
                    py += 1
                }
            } else {
                if (__pytra_int(kind) == __pytra_int(Int64(1))) {
                    var ph: Int64 = __pytra_len(r_pentomino)
                    var py = __pytra_int(Int64(0))
                    while (py < __pytra_int(ph)) {
                        var pw: Int64 = __pytra_len(__pytra_as_list(__pytra_getIndex(r_pentomino, py)))
                        var px = __pytra_int(Int64(0))
                        while (px < __pytra_int(pw)) {
                            if (__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(r_pentomino, py)), px)) == __pytra_int(Int64(1))) {
                                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, ((gy + py) % h))), ((gx + px) % w), Int64(1))
                            }
                            px += 1
                        }
                        py += 1
                    }
                } else {
                    var ph: Int64 = __pytra_len(lwss)
                    var py = __pytra_int(Int64(0))
                    while (py < __pytra_int(ph)) {
                        var pw: Int64 = __pytra_len(__pytra_as_list(__pytra_getIndex(lwss, py)))
                        var px = __pytra_int(Int64(0))
                        while (px < __pytra_int(pw)) {
                            if (__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(lwss, py)), px)) == __pytra_int(Int64(1))) {
                                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, ((gy + py) % h))), ((gx + px) % w), Int64(1))
                            }
                            px += 1
                        }
                        py += 1
                    }
                }
            }
            gx += __step_3
        }
        gy += __step_2
    }
    var frames: [Any] = __pytra_as_list([])
    var __loop_10 = __pytra_int(Int64(0))
    while (__loop_10 < __pytra_int(steps)) {
        frames.append(render(grid, w, h, cell))
        grid = __pytra_as_list(next_state(grid, w, h))
        __loop_10 += 1
    }
    __pytra_noop(out_path, (w * cell), (h * cell), frames, [])
    var elapsed: Double = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_07_game_of_life_loop()
    }
}
